fdbclient/GlobalConfig.actor.cpp

```diff
@@ -186,52 +191,61 @@ ACTOR Future<Void> GlobalConfig::updater(GlobalConfig* self, const ClientDBInfo*
 			if (self->initialized.canBeSet()) {
 				wait(self->cx->onConnected());
 
-				wait(self->refresh(self, -1));
+				Version version = wait(self->refresh(self, -1, 0));
+				self->lastUpdate = version;
+
+				self->cx->addref();
 				self->initialized.send(Void());
+				self->cx->delref();
 			}
 
 			loop {
 				try {
-					wait(self->dbInfoChanged.onTrigger());
-
-					auto& history = dbInfo->history;
-					if (history.size() == 0) {
-						continue;
-					}
-
-					if (self->lastUpdate < history[0].version) {
-						// This process missed too many global configuration
-						// history updates or the protocol version changed, so it
-						// must re-read the entire configuration range.
-						wait(self->refresh(self, history.back().version));
-						if (dbInfo->history.size() > 0) {
-							self->lastUpdate = dbInfo->history.back().version;
-						}
-					} else {
-						// Apply history in order, from lowest version to highest
-						// version. Mutation history should already be stored in
-						// ascending version order.
-						for (const auto& vh : history) {
-							if (vh.version <= self->lastUpdate) {
-								continue; // already applied this mutation
+					// run one iteration at the beginning
+					wait(delay(0));
+					if (dbInfo->history.size() > 0) {
+						if (self->lastUpdate < dbInfo->history[0].version) {
+							// This process missed too many global configuration
+							// history updates or the protocol version changed, so it
+							// must re-read the entire configuration range.
+							Version version =
+							    wait(self->refresh(self, self->lastUpdate, dbInfo->history.back().version));
+							self->lastUpdate = version;
+							// DBInfo could have changed after the wait. If
+							// changes are present, re-run the loop to make
```

Comment:
I am confused. If `dbInfo->history[0].version != std::numeric_limits<Version>::max()`, why can't we apply the changes? If we want to be sure of DBInfo changes, could store `self->dbInfoChanged.onTrigger()` in a Future object and check if the future is ready here.

Reply:
i discussed this with @sfc-gh-ljoswiak , we want to hit `self->dbInfoChanged.onTrigger()` when it is int_max, can you please explain a bit more for jingyu's question? @sfc-gh-ljoswiak

Reply:
i believe local copy is updated in `refresh()` method, my understanding was that, in the case of `int_max` , we know we have updated copy, and now we should just go `wait(self->dbInfoChanged.onTrigger());` for the next change.

Reply:
Yes, `refresh` will update the local state. Since `int_max` is a special case, if we do a `continue` and immediately run the loop again we will get into an infinite loop with no `wait`s, since the GRV proxy will continue to reply to the `GlobalConfigRefreshRequest` right away.

Reply:
ok. i get it now.